File: go/14LongestCommonPrefix.go
package main

import (
	"fmt"
)

func longestCommonPrefix(strs []string) string {
	lenStrs := len(strs)
	if lenStrs == 0 {
		return ""
	}
	if lenStrs == 1 {
		return strs[0]
	}
	if lenStrs > 2 {
		return longestCommonPrefix([]string{longestCommonPrefix(strs[:lenStrs-1]), strs[lenStrs-1]})
	}

	output := []byte{}

	for i := 0; i < len(strs[0]) && i < len(strs[1]); i++ {
		if strs[0][i] == strs[1][i] {
			output = append(output, byte(strs[0][i]))
		} else {
			break
		}
	}

	return string(output)
}

func main() {
	output := longestCommonPrefix([]string{"flower", "flow", "flight"})
	output1 := longestCommonPrefix([]string{"dog", "racecar", "car"})
	output2 := longestCommonPrefix([]string{"cir", "car"})

	fmt.Println("\n******** output *********\n", output, output1, output2)
}

File: cpp/14LongestCommonPrefix.cpp
#include <iostream>
#include <set>
#include <vector>
#include <iterator>
#include <utility>
#include <cstdlib>
#include <limits>
#include <functional>
#include <stack>
#include <unordered_set>
#include <unordered_map>
#include <queue>
#include <algorithm>
#include <cstdint>

using namespace std;

template<class Container>
void printContainer(const Container& container) {
//    return;

    if (container.size() == 0) {
        cout << "[]" << endl;
        return;
    }

    for (auto it = container.begin(); it != container.end(); it++) {
        if (it == container.begin()) {
            cout << "[";
        }
        cout << *it;
        if (it == prev(container.end())) {
            cout << "]" << endl;
        } else {
            cout << ", ";
        }
    }
}

template<class T>
void print2DimVec(const vector<vector<T>>& container) {
//    return;

    if (container.size() == 0) {
        cout << "[]" << endl;
        return;
    }
    for (auto it = container.begin(); it != container.end(); it++) {
        if (it == container.begin()) {
            cout << "[" << endl;
        }
        printContainer(*it);
        if (it == prev(container.end())) {
            cout << "]" << endl;
        }
    }
}

template<class T>
void printMap(const T& container) {
//    return;

    cout << "size " << container.size() << endl;
    if (container.size() == 0) {
        cout << "[]" << endl;
        return;
    }
    for (auto it = container.begin(); it != container.end(); it++) {
        if (it == container.begin()) {
            cout << "[ ";
        }

        cout << "[" << it->first << ", " << it->second << "]";

        if (next(it) == container.end()) {
            cout << " ]" << endl;
        } else {
            cout << ", ";
        }
    }
}

class Solution {
public:
    string longestCommonPrefix(vector<string>& strs) {
        if (strs.empty()) {
            return "";
        }
        int i = 0;
        while (true) {
            int c = 0;
            for (const auto& s : strs) {
                if (i == s.size()) {
                    return strs[0].substr(0, i);
                }

                if (c == 0) {
                    c = static_cast<int>(s[i]);
                } else if (c != static_cast<int>(s[i])) {
                    return strs[0].substr(0, i);
                }
            }
            i++;
        }

        return strs[0].substr(0, i);
    }
};

int main(){
    Solution s;

    auto input = vector<string>{"flower", "flow", "flight"};
	auto output = s.longestCommonPrefix(input);

    auto input1 = vector<string>{"dog", "racecar", "car"};
	auto output1 = s.longestCommonPrefix(input1);

    auto input2 = vector<string>{"cir", "car"};
	auto output2 = s.longestCommonPrefix(input2);

    cout << "*************output*************" << endl;
    cout << output << endl;
    cout << output1 << endl;
    cout << output2 << endl;

    return 0;
}

